Add stbt-docker: Used when video-capture hardware isn't required

wmanley opened the pull request:
Runs the stbt command in a docker container, set-up like an stb-tester ONE but without video capture or IR control.  This is intended to be used as the "stbt" command on MacOS and Windows allowing people to do local development and testing (pylint, running selftests, etc.) when video-capture is not needed.

Usage:

    $ stbt-docker lint tests/tests.py
    ********** module tests
    E: 22, 4: No name '_load_template' in module 'stbt' (no-name-in-module)

    $ stbt-docker exec python -m doctest tests/tests.py
    ...

    $ stbt-docker exec
    Python 2.7.6 (default, Jun 22 2015, 17:58:13)
    [GCC 4.8.2] on linux2
    Type "help", "copyright", "credits" or "license" for more information.
    >>>

Environment variables:

* `DOCKER` - The location of the `docker` binary defaults to "`docker`"
* `DOCKER_OPTS` - provide additional arguments to `docker run`.  For instance `DOCKER_OPTS="--memory=2g"` will limit stbt to using 2GB of memory.  Shell escaping is used to allow specifiying arguments with spaces in.

This program is built with portability in mind so it should run on Mac OS and Windows.  The only dependencies are Python and docker.  It is also self-contained and relocatable so it can be deployed as a single file with no dependency on anything else in stbt.

~~In the future it might be worth re-writing this in Go so we can drop the Python dependency too.~~

This was inspired by writing the blog post on auto-regression tests.  I was concerned that we were recommending an approach without making it easy for our users to execute on it.

This is a generalisation of a `docker-env` script I've been using to create a more consistent selftest environment.

This could also be used to make virtual-stb merge-able sooner (#174).

TODO:

- [x] Write tests
- [x] ~~Merge #303 so we can run the tests on Travis~~ - I'm skipping the tests if docker is not available instead.  It's better to decouple these sorts of things.
- [x] Figure out permissions and how volumes work on MacOS and Windows
- [x] Figure out how it works if the outer UID is not 1000
- [x] Automated test for UID/GID switching?
- [x] ~~Maybe it would be better to `git clone` into the container rather than using host volumes?~~ - Nope, see comment below
- [x] Write release note
- [x] Change command-line API as per [this comment](#issuecomment-191826936).
- [x] Change command-line API when no arguments as per [this comment](#issuecomment-215490797).
- [x] Decide which URL to use for upgrades, as per [this comment thread](#discussion-diff-54333636) *(Now that it's going into its own repo, the URL will be `raw.github.com/...`)*
- [x] Add support for Windows and Mac OS X with docker-toolbox.
- [ ] Remove it from `stbt.sh.in` -- we'll use it with `$test_pack/stbt-docker` not `stbt docker`. This is because stbt-docker will have a different release cycle than stb-tester.
    - [ ] Check all the documentation to make sure it isn't saying `stbt docker`.
- [x] Move to seperate repo

wmanley commented:
> Maybe it would be better to git clone into the container rather than using host volumes?

Nah, I don't think so.  The disadvantage of volumes as implemented here is that the docker daemon has to be running on the same machine as `stbt-docker`.  In the future we can think about adding other options for getting the test-pack into the container such that it works remotely.  Alternatively we could look into using one of the fancy [volume plugins] to allow mounting using sshfs or similar, but I don't think we need to worry about this just yet.

[volume plugins]: https://docs.docker.com/extend/plugins_volume/

drothlis commented:
Will have to check how volumes work with docker on OS X. In the boot2docker days it certainly wasn't trivial sharing files between the host & the container. IIRC there's now a docker native replacement for boot2docker but we'll still need to check how volumes work.

wmanley commented:
> Will have to check how volumes work with docker on OS X

My understanding is that everything under /Users is available to be mounted.  From the [docker documentation](https://docs.docker.com/userguide/dockervolumes/):

> If you are using Docker Machine on Mac or Windows, your Docker daemon only has limited access to your OS X/Windows filesystem. Docker Machine tries to auto-share your /Users (OS X) or C:\Users (Windows) directory - and so you can mount files or directories using docker run -v /Users/<path>:/<container path> ... (OS X) or docker run -v /c/Users/<path>:/<container path ... (Windows). All other paths come from your virtual machine’s filesystem.

wmanley commented:
CI failure is unrelated:

    TypeError: Expected a Gst.Structure, but got Structure

Hopefully #303 will fix some of this CI unreliability.

wmanley commented:
Ok, I've pushed a change so the UID and GID are matched between inside and outside the container.  I've tested this by running the existing tests and also:

    $ sudo -u steam bash
    $ id
    uid=1001(steam) gid=1001(steam) groups=1001(steam),128(docker)
    $ ./stbt-docker exec id
    uid=1001(stb-tester) gid=1001(stb-tester) groups=1001(stb-tester)
    $ ./stbt-docker exec ls /var/lib/stbt -la
    total 32
    drwxr-xr-x  5 stb-tester 1001 4096 Sep 18 14:06 .
    drwxr-xr-x 28 root       root 4096 Jul 31 16:15 ..
    -rw-r--r--  1 stb-tester 1001  220 Jul 31 16:13 .bash_logout
    -rw-r--r--  1 stb-tester 1001 3637 Jul 31 16:13 .bashrc
    drwxr-xr-x  2 stb-tester 1001 4096 Sep 17 15:01 .cache
    -rw-r--r--  1 stb-tester 1001  675 Jul 31 16:13 .profile
    drwxr-xr-x  2 stb-tester 1001 4096 Sep  8 16:43 results
    drwxr-xr-x  4 stb-tester 1001 4096 Sep 18 12:48 test-pack

But I can't think of a way of writing an automated test for this.

drothlis commented:
When I run this I'm getting `/bin/sh: 1: config/setup/setup: Text file busy` during the `docker build` step:

```
$ ./stbt-docker exec ls
Sending build context to Docker daemon 10.24 kB
Sending build context to Docker daemon 
Step 0 : FROM stbtester/stbt-one-test-pack-base-23:latest
 ---> f1729dcb11e3
Step 1 : RUN mkdir -p /var/lib/stbt/test-pack/config/setup
 ---> Using cache
 ---> 6823d1f62311
Step 2 : ADD setup /var/lib/stbt/test-pack/config/setup/setup
 ---> Using cache
 ---> 6632416601b8
Step 3 : RUN cd /var/lib/stbt/test-pack &&     sudo chown stb-tester:stb-tester config/setup/setup &&     chmod a+x config/setup/setup &&     config/setup/setup
 ---> Running in f2d3bdccb95a
/bin/sh: 1: config/setup/setup: Text file busy
INFO[0000] The command [/bin/sh -c cd /var/lib/stbt/test-pack &&     sudo chown stb-tester:stb-tester config/setup/setup &&     chmod a+x config/setup/setup &&     config/setup/setup] returned a non-zero code: 2 
Traceback (most recent call last):
  File "./stbt-docker", line 184, in <module>
    sys.exit(main(sys.argv))
  File "./stbt-docker", line 139, in main
    docker_image = build_docker_image(root)
  File "./stbt-docker", line 125, in build_docker_image
    stdout=sys.stderr, stdin=f)
  File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['docker', 'build', '-t', 'stb-tester-test-pack-fcc8df490ab5d4534c71402f4a5d53e9', '-']' returned non-zero exit status 1
```

The following patch fixed it, or at least changed the time enough to make it work once:

```
diff --git a/stbt-docker b/stbt-docker
index 138baab..68b5152 100755
--- a/stbt-docker
+++ b/stbt-docker
@@ -72,10 +72,9 @@ FROM {base_image}
 
 RUN mkdir -p /var/lib/stbt/test-pack/config/setup
 ADD setup /var/lib/stbt/test-pack/config/setup/setup
-RUN cd /var/lib/stbt/test-pack && \\
-    sudo chown stb-tester:stb-tester config/setup/setup && \\
-    chmod a+x config/setup/setup && \\
-    config/setup/setup
+RUN sudo chown stb-tester:stb-tester /var/lib/stbt/test-pack/config/setup/setup
+RUN chmod a+x /var/lib/stbt/test-pack/config/setup/setup
+RUN /var/lib/stbt/test-pack/config/setup/setup
 """
 ```

drothlis commented on stbt-docker:
```diff
@@ -0,0 +1,184 @@
+         '&& sed -i s/stb-tester:x:1000/stb-tester:x:%i/ /etc/group '
+         '&& cd %s && sudo -u stb-tester -E -- "$0" "$@"'
+         % (os.getuid(), os.getgid(), os.getuid(), os.getgid(), os.getgid(),
+            pipes.quote(innerdir))] + args)
```
`os.getuid` etc. won't be portable to Windows.

We can still merge this, because it's useful for things like running an ipython notebook will all your test-pack's dependencies installed. We can figure out the portability later.

wmanley replied:
> `os.getuid` etc. won't be portable to Windows.

And it would be inappropriate for MacOS X with boot2docker.  The UIDs inside the VM will have no relation to on the MacOS host.

> We can figure out the portability later

Agreed.

wmanley commented:
> When I run this I'm getting /bin/sh: 1: config/setup/setup: Text file busy during the docker build step:

Seems like docker/docker#9547.

wmanley commented:
I'll add this to [stb-tester/stb-tester-test-pack](https://github.com/stb-tester/stb-tester-test-pack) once it's merged.

drothlis commented on stbt-docker:
```diff
@@ -0,0 +1,246 @@
+            "This version of stbt-docker does not support stbt_version %r.  "
+            "Update to a newer version with:\n"
+            "\n"
+            "    curl -f -o %s https://stb-tester.com/stbt-docker\n")
```
Perhaps this URL should be https://raw.github.com/stb-tester/stb-tester-test-pack/master/stbt-docker so that we don't have to remember to update yet another place with each release?

drothlis replied:
Also this is missing both string substitution arguments.

drothlis replied:
Also, we don't want to print this message if you're using the stbt-docker provided by the official stb-tester ubuntu package.

wmanley replied:
The reason I did it like this is that I didn't want to depend on the location changing.  I was thinking that we could set up a redirect.  But yeah, it seems that we're unlikely to move away from github imminently so I'll change this.  I'm not sure if we want to link to the stb-tester-test-pack repo or the stb-tester repo though.

drothlis replied:
>  I'm not sure if we want to link to the stb-tester-test-pack repo or the stb-tester repo though.

The advantage of pointing at the stb-tester-test-pack repo is that it will only get updated with each release, whereas with the stb-tester repo you could get a version that isn't ready for release.

wmanley replied:
> https://raw.github.com/stb-tester/stb-tester-test-pack/master/stbt-docker

Hmm, so if we were to distribute it in stb-tester-test-pack would it live in the root like this?  Or `tools/stbt-docker` or `tools/stbt-docker/stbt-docker`?

Or as a separate repo entirely given that it has a different release cadence?

drothlis commented on stbt-docker:
```diff
@@ -0,0 +1,246 @@
+            "This command must be run within a test pack.  Couldn't find a "
+            ".stbt.conf in this or any parent directory.\n"
+            "\n"
+            "To create a new test pack run:\n"
```
Perhaps we should have a web page or wiki page documenting the structure of a test-pack, and link to that -- allowing us to update the instructions independently of the stbt-docker you have installed.

wmanley replied:
While a page describing the structure of a test-pack would be useful as a reference I wouldn't want to link to that in the instructions on how to get started.  I feel like it would be like linking to [man gitrepository-layout] when all people want to know is `git init`.  I don't particularly want to have instructions to update, although I agree that these are instructions I think they're pretty likely to be stable unless we move away from github.

[man gitrepository-layout]: http://git-scm.com/docs/gitrepository-layout

drothlis commented on stbt.sh.in:
```diff
@@ -44,7 +44,7 @@ case "$1" in
     -v|--version)
         echo "stb-tester $STBT_VERSION"; exit 0;;
-    run|record|batch|config|control|lint|power|screenshot|templatematch|tv)
+    run|record|batch|config|control|docker|lint|power|screenshot|templatematch|tv)
```
Would we want users to say `stbt docker run ...` or `stbt --docker run`? Seeing it's supposed to behave just like `stbt`...

On the other hand, `stbt docker --help` does make more sense than `stbt --docker --help`.

drothlis commented:
I have a few usability & documentation fixes I want to make (e.g. running `stbt-docker --help` launches into a scary docker build process).

drothlis commented:
Hmmm... there are 2 different uses: One intended to transparently replace `stbt` and one intended as a sub-command of stbt (`stbt docker ...`).

For the first case it would be good if `stbt-docker --help` didn't have to build & run the docker container.

If we re-implemented `stbt.sh` in python, `stbt` and `stbt-docker` could re-use the same argparse help message, perhaps through copy-and-paste or perhaps by merging `stbt` and `stbt-docker` into a single file.

wmanley commented:
Ok, I've just spoken to Dave and we agree that the UI isn't quite right here.  `stbt-docker` should just execute a program rather than pretending to be `stbt`.  So instead of:

    stbt-docker lint

we will write

    stbt-docker stbt lint

and instead of

    stbt-docker exec make check

we will write

    stbt-docker make check

wmanley commented:
@drothlis: All the TODO list items are ticked now. What do you reckon?

wmanley commented:
CI failure is unrelated

drothlis commented:
The STBT_VERSION_WHITELIST error message still points to `http://stb-tester.com/stbt-docker` whereas we seemed to have [agreed](#discussion-diff-54333636) to change it to point at github?

Also [this comment](#discussion_r54333643) still needs addressing:

> this is missing both string substitution arguments.

drothlis commented:
@wmanley & I have agreed that the default behaviour of `./stbt-docker` (without arguments) is to print a usage message rather than run an interactive python interpreter.

drothlis commented:
I've pushed a couple of commits with documentation improvements.

drothlis commented:
I've pushed a commit that removes the special behaviour when you don't give any arguments. Now the only thing left on the todo list is the URL that we print in [this error message](https://github.com/stb-tester/stb-tester/pull/304/files#diff-0dc4705ad20f1d65eb4b0a6871e8a8bcR223).

wmanley commented:
I changed my mind.  I think MacOS and Windows support is too important to leave out now that we've got `stbt auto-selftest` on master.  I've implemented support.  I'll have to write a blog post about this at some point with installation instructions.

wmanley commented:
Perhaps this doesn't belong in the stb-tester repo at all:

* The release cadence is potentially different
* It has to be standalone and can't depend on any code in stb-tester to function
* The licence should be MIT or something so people aren't scared to include it in their test-packs verbatim
* It shouldn't be distributed as a part of stb-tester
 
I've added a TODO list item to create a different repo for this.

wmanley commented:
See [stb-tester/stbt-docker](https://github.com/stb-tester/stbt-docker.git) for its new home.

drothlis commented:
This has been moved to https://github.com/stb-tester/stbt-docker

Closing & removing from v25 milestone.
